File: backend/src/main/java/backend/entity/taskcard/Task.java
package backend.entity.taskcard;

import jakarta.persistence.*;
import lombok.Getter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Entity
@Table(name = "task")
@Getter
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long taskId;


    @Column(nullable = false)
    private Long projectId;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(length = 2000)
    private String description;

    @Column(nullable = false, length = 30)
    @Enumerated(EnumType.STRING)
    private TaskStatus status; // TODO, IN_PROGRESS, DONE, DELETED

    // tags를 "tag1,tag2" 문자열로 저장 (가장 단순)
    @Column(length = 1000)
    private String tagsCsv;

    @Lob
    @Column(columnDefinition = "TEXT")
    private String notes; // 마크다운

    private Long assigneeId; //담당자(Id)

    private LocalDateTime createdDate;
    private LocalDateTime updatedDate;
    private LocalDateTime dueDate;
//    private TaskDeleteReason taskDeleteReason;

    protected Task() {}

    public Task(Long projectId, String title, String description, TaskStatus status,
                String tagsCsv, String notes, Long assigneeId, LocalDateTime dueDate,
                LocalDateTime createdDate, LocalDateTime updatedDate) {
        this.projectId = projectId;
        this.title = title;
        this.description = description;
        this.status = status;
        this.tagsCsv = tagsCsv;
        this.notes = notes;
        this.assigneeId = assigneeId;
        this.dueDate = dueDate;
        this.createdDate = createdDate;
        this.updatedDate = updatedDate;
    }


    // ===== update methods =====
    public void updateBasic(String title, String description, TaskStatus status, String tagsCsv, String notes,
                            Long assigneeId, LocalDateTime dueDate, LocalDateTime updatedAt) {
        if (title != null) this.title = title;
        if (description != null) this.description = description;
        if (status != null) this.status = status;
        if (tagsCsv != null) this.tagsCsv = tagsCsv;
        if (notes != null) this.notes = notes;
        if (assigneeId != null) this.assigneeId = assigneeId;
        if (dueDate != null) this.dueDate = dueDate;

        this.updatedDate = updatedAt;
    }

    // tags 변환 헬퍼(원하면 Mapper로 빼도 됨)
    public static String toCsv(List<String> tags) {
        if (tags == null || tags.isEmpty()) return "";
        // 공백 제거 + 빈 문자열 제거
        List<String> cleaned = new ArrayList<>();
        for (String t : tags) {
            if (t == null) continue;
            String s = t.trim();
            if (!s.isEmpty()) cleaned.add(s);
        }
        return String.join(",", cleaned);
    }

    public static List<String> fromCsv(String csv) {
        if (csv == null || csv.isBlank()) return List.of();
        return Arrays.stream(csv.split(","))
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .toList();
    }
    public void changeAssignee(Long assigneeId, LocalDateTime updatedDate) {
        this.assigneeId = assigneeId;
        this.updatedDate = updatedDate;
    }

    public void restoreTask(TaskStatus status, LocalDateTime updatedDate) {
        if (this.status != TaskStatus.DELETED) {
            throw new IllegalStateException("Deleted task만 복구 가능");
        }
        this.status = status;
        this.updatedDate = updatedDate;
    }
}

File: frontend/src/features/tasks/types/task.js
/**
 * Task Entity 속성 정의
 * @typedef {Object} TaskData
 * @property {number} taskId
 * @property {string} title
 * @property {string | null} description
 * @property {'TODO' | 'IN_PROGRESS' | 'DONE' | 'DELETED'} status
 * @property {string[] | null} tags
 * @property {string | null} notes
 * @property {number} assigneeId                    // memberId는 없는데 그럼 username?
 * @property {string} assigneeName                  // nickname?
 * @property {string | null} assigneeProfileImage   // 현재 유저 엔티티에 없음
 * @property {Date | null} startDate                // 필요한데 없음
 * @property {Date | null} dueDate
 * @property {Date} createdDate
 * @property {Date} updatedDate
 */

/**
 * 작업 카드 요청 DTO (클라이언트 요청 형식)
 * @see TaskRequestDTO.java
 * @see ../../../../../backend/src/main/java/backend/dto/taskcard/TaskRequestDTO.java
 * @typedef {Pick<TaskData, 'title' | 'description' | 'status' | 'tags' | 'notes' | 'assigneeId' | 'startDate' | 'dueDate'>} TaskRequestDTO
 */

/**
 * [목록 조회용] 작업 요약 정보 DTO
 * @typedef {Object} TaskSummaryDTO
 * @property {TaskData['taskId']} taskId - 작업 카드 고유 ID
 * @property {TaskData['title']} title - 작업 카드 제목
 * @property {TaskData['description']} [description] - 작업 설명 (Optional)
 * @property {TaskData['status']} status - 작업 상태
 * @property {TaskData['tags']} tags - 태그 목록 (예: ["기획", "디자인"])
 * @property {string} [createdAt] - 생성일 (ISO Date String)
 * @property {string} [updatedAt] - 수정일 (ISO Date String)
 * @property {string} [dueDate] - 마감일 (ISO Date String)
 * @property {TaskData['assigneeId']} [assigneeId] - 담당자 ID
 * @property {TaskData['assigneeName']} [assigneeName] - 담당자 이름 (백엔드 연동 시 채워짐)
 */

export {}; // 모듈로 인식되도록 export
